lib/rdoc/server.rb

```diff
@@ -0,0 +1,182 @@
+      res.content_type = 'application/javascript'
+    else
+      show_documentation req, res
+    end
+  rescue WEBrick::HTTPStatus::NotFound => e
+    not_found @generator, req, res, e.message
+  rescue WEBrick::HTTPStatus::Status
+    raise
+  rescue => e
+    $stderr.puts e.full_message
+    error e, req, res
+  end
+
+  private
+
+  def asset generator_name, req, res
+    asset_dir = @asset_dirs[generator_name]
+
+    asset_path = File.join asset_dir, req.path
+
+    res.body = File.read asset_path
+
+    res.content_type = case req.path
+                       when /\.css\z/ then 'text/css'
+                       when /\.js\z/  then 'application/javascript'
+                       else                'application/octet-stream'
+                       end
+  end
+
+  def error exception, req, res
+    backtrace = exception.backtrace.join "\n"
+
+    res.content_type = 'text/html'
+    res.status = 500
+    res.body = <<-BODY
+<!DOCTYPE html>
+<html>
+<head>
+<meta content="text/html; charset=UTF-8" http-equiv="Content-Type">
+
+<title>Error - #{ERB::Util.html_escape exception.class}</title>
+
+<link type="text/css" media="screen" href="/css/rdoc.css" rel="stylesheet">
+</head>
+<body>
+<h1>Error</h1>
+
+<p>While processing <code>#{ERB::Util.html_escape req.request_uri}</code> the
+RDoc (#{ERB::Util.html_escape RDoc::VERSION}) server has encountered a
+<code>#{ERB::Util.html_escape exception.class}</code>
+exception:
+
+<pre>#{ERB::Util.html_escape exception.message}</pre>
+
+<p>Please report this to the
+<a href="https://github.com/ruby/rdoc/issues">RDoc issues tracker</a>.  Please
+include the RDoc version, the URI above and exception class, message and
+backtrace.  If you're viewing a gem's documentation, include the gem name and
+version.  If you're viewing Ruby's documentation, include the version of ruby.
+
+<p>Backtrace:
+
+<pre>#{ERB::Util.html_escape backtrace}</pre>
+
+</body>
+</html>
+    BODY
+  end
+
+  def not_found generator, req, res, message = nil
+    message ||= "The page <kbd>#{ERB::Util.h req.path}</kbd> was not found"
+    res.body = generator.generate_servlet_not_found message
+    res.status = 404
+  end
+
+  PAGE_NAME_SUB_REGEXP = /_([^_]*)\z/
+
+  def show_documentation req, res
+    store = @rdoc.store
+    # Clear all the previous data
+    store.classes_hash.clear
+    store.modules_hash.clear
+    store.files_hash.clear
+
+    # RDoc instance use last_modified list to avoid reparsing files
+    # We need to clear it to force reparsing
+    @rdoc.last_modified.clear
+
+    # Generator copies data from store to instance variables
+    # We need to clear it before regenerating the documentation
+    @generator.clear_data
+
+    # Reparse the files
+    @rdoc.parse_files(@rdoc.options.files)
+
+    # Update the data from store
+    @generator.update_data_from_store
+
+    # Regenerate the documentation and asserts
+    @generator.generate(server_mode: true)
```

Comment:
Through implementing this feature I learned a lot about how data are passed around, which sometimes simply duplicate in multiple classes. And I plan to do some refactor around them.